kvdb/postgres/db.go

```diff
@@ -0,0 +1,238 @@
+// config. If postgres connection cannot be estabished, then returns error.
+func newPostgresBackend(ctx context.Context, config *Config, prefix string) (
+	*db, error) {
+
+	if prefix == "" {
+		return nil, errors.New("empty postgres prefix")
+	}
+
+	dbConn, err := sql.Open("pgx", config.Dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	// Log connection parameters.
+	timeoutString := "none"
+	if config.Timeout != 0 {
+		timeoutString = config.Timeout.String()
+	}
+	log.Infof("Connected to Postgres: dsn=%v, prefix=%v, timeout=%v",
+		config.Dsn, prefix, timeoutString)
+
+	// Compose system table names.
+	hexTableName := fmt.Sprintf(
+		"%s%s_%s", prefix, systemTablePrefixExtension, hexTableId,
+	)
+
+	sequenceTableName := fmt.Sprintf(
+		"%s%s_%s", prefix, systemTablePrefixExtension, sequenceTableId,
+	)
+
+	hexCreateTableSql := getCreateTableSql(hexTableName)
+
+	// Create schema and system tables.
+	_, err = dbConn.ExecContext(ctx, `
+	CREATE SCHEMA IF NOT EXISTS public;
+
+	CREATE TABLE IF NOT EXISTS public.`+sequenceTableName+`
+	(
+		table_name TEXT NOT NULL PRIMARY KEY,
+		sequence BIGINT
+	);
+	`+hexCreateTableSql)
+	if err != nil {
+		_ = dbConn.Close()
+
+		return nil, err
+	}
+
+	backend := &db{
+		cfg:               config,
+		prefix:            prefix,
+		ctx:               ctx,
+		db:                dbConn,
+		hexTableName:      hexTableName,
+		sequenceTableName: sequenceTableName,
+	}
+
+	return backend, nil
+}
+
+// getTimeoutCtx gets a timeout context for database requests.
+func (db *db) getTimeoutCtx() (context.Context, func()) {
+	if db.cfg.Timeout == time.Duration(0) {
+		return db.ctx, func() {}
+	}
+
+	return context.WithTimeout(db.ctx, db.cfg.Timeout)
+}
+
+// getPrefixedTableName returns a table name for this prefix (namespace).
+func (db *db) getPrefixedTableName(table string) string {
+	return fmt.Sprintf("%s_%s", db.prefix, table)
+}
+
+// catchPanic executes the specified function. If a panic occurs, it is returned
+// as an error value.
+func catchPanic(f func() error) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+			log.Criticalf("Caught unhandled error: %v", err)
+		}
+	}()
+
+	err = f()
+
+	return
+}
+
+// View opens a database read transaction and executes the function f with the
+// transaction passed as a parameter. After f exits, the transaction is rolled
+// back. If f errors, its error is returned, not a rollback error (if any
+// occur). The passed reset function is called before the start of the
+// transaction and can be used to reset intermediate state. As callers may
+// expect retries of the f closure (depending on the database backend used), the
+// reset function will be called before each retry respectively.
+func (db *db) View(f func(tx walletdb.ReadTx) error, reset func()) error {
+	reset()
+
+	tx, err := newReadWriteTx(db, true)
+	if err != nil {
+		return err
+	}
+
+	err = catchPanic(func() error { return f(tx) })
+	if err != nil {
+		_ = tx.Rollback()
```

Comment:
Should we log the rollback error? That way you'd at least see how many transactions were affected by the panic/error. Though because of the lock it should only ever be one at a time... So maybe not that useful at all.

Reply:
Doesn't hurt to log it, added.